### DISCUSSION TITLE
server.root_url not honored

### DISCUSSION THREAD

**Discussion by joerglemke** [gapitio/gapit-htmlgraphics-panel#19] [General] [answered]
Hello folks,
I tried to use your plugin because it sounds promising to achieve what I want. However it fails on load.

In my installation Grafana is running on a custom root URL, i.e. server.root_url is set.
This causes HTML Graphics plugin to fail on load of loader.js and a font file, because the root path is not used in the request.

I was successful when editing the module.js inside the running docker container by prefixing all occurences of "/public/plugins/gapit-htmlgraphics-panel" with the root path. One for loader.js, some more for the fonts.

I found this configuration at https://github.com/gapitio/gapit-htmlgraphics-panel/blob/main/src/components/TextEditor/index.tsx line 12, where the path is hard coded. I believe the Grafana configuration of root_url should be prepended here.
However, I could not find where the font paths are built in the source code.

**Comment by ZuperZee** [comment]
Hello @joerglemke.
Thank you for checking out the plugin.

The fonts are loaded from the scss file <https://github.com/gapitio/gapit-htmlgraphics-panel/blob/main/src/fonts.scss>.
The fonts should be relative to the plugin folder (/public/plugins/gapit-htmlgraphics-panel/fonts/OpenSans-Regular.woff2).

I'll try to fix the issues with the paths.

**Reply by joerglemke** [reply]
Thanks @ZuperZee for looking into the issue.
Regarding the fonts. The module.js file (output from the typescript compilation) has 4 occurences of the font paths starting with "/public/plugins/gapit-htmlgraphics-panel". This is also the absolute path attempted by the browser to load the font.
Although the fonts are indeed in the plugin folder on the file system, due to the root_url configuration ("/my/special/prefix") the font should be loaded from the URL path "/my/special/prefix/public/plugins/gapit-htmlgraphics-panel/fonts/...."
fonts.scss only mentions relative paths. So, somehow the path prefix "/public/plugins/gapit-htmlgraphics-panel" is added during compilation.

**Reply by ZuperZee** [reply]
@joerglemke, I'm having some problems recreating the problem.
Changed the config like this.
![image](https://user-images.githubusercontent.com/39485579/113697581-a971b780-96d3-11eb-908a-324e88292081.png)
The url seems correct.
![image](https://user-images.githubusercontent.com/39485579/113697698-d7ef9280-96d3-11eb-98e7-594fc08323f7.png)
Everything seems to load properly.
![image](https://user-images.githubusercontent.com/39485579/113697797-fa81ab80-96d3-11eb-9542-8782b238d65c.png)

Could you send a minimal config of your setup? (dockerfile, docker-compose, grafana config)

**Reply by joerglemke** [reply]
@ZuperZee I sent an excerpt from my config by mail. Have you received it?
In the meantime I have checked some possible changes and combinations of it.

First: my config only had "server.root_url" set. "server.serve_from_sub_path" was not set and thus false by default. This was ok, because my reverse proxy was configured for stripping the path prefix. Everything was working fine for a long time with other plugins. Setting server.serve_from_sub_path=true requires to remove that stripping from the reverse proxy. But then everything still works as before. In both cases however, the above mentioned problems with the Monace editor loader persist.

I edited the module.js (the file created during plugin compilation) in place on the running installation and found, that there are 5 occurences of the path prefix "/public/plugins/gapit-htmlgraphics-panel": 1 for the loader script and 4 for the fonts. Then I removed the leading slash in all places, i.e. "public/plugins/gapit-htmlgraphics-panel..." and voila the problems are gone, because the path is relative then. This should also work when the root_url path is empty.

For the URL of the loader script this can be changed easily inside the source file TextEditor/index.tsx: just remove the slash there.

For the fonts however the path prefix seems to be generated by the grafana-toolkit. Until now I did not look into that any deeper yet.

Actually, missing out the fonts does not seem to cause much harm. Getting loader.js to be loaded however is essential for the editors to work.

So (just) my suggestion, if ok with you: let's try to remove that leading slash from the loader path. What do you think?

**Reply by ZuperZee** [reply]
I got your email @joerglemke.

It should be fine to remove the leading slash from the monaco editor loader path.

I'll look into how to change the font path before making a new release. (Hopefully by the end of the day)

**Reply by ZuperZee** [reply]
Looks like the fonts will be harder to get working without making changes to the @grafana/toolkit files.

Created an issue on their repo <https://github.com/grafana/grafana/issues/32750>.
Hopefully someone with more knowledge about it can help out.

**Reply by ZuperZee** [reply]
A pull request has been made to the plugin repo <https://github.com/grafana/grafana-plugin-repository/pull/917>.
When it gets merged the release v1.3.2 will be availbe through grafana with the normal `gapit-htmlgraphics-panel`

The new release is also available with `https://github.com/gapitio/gapit-htmlgraphics-panel/releases/download/v1.3.2/gapit-htmlgraphics-panel-1.3.2.zip;gapit-htmlgraphics-panel`

**Reply by joerglemke** [reply]
Thanks, the version from Github is working fine.

Font loading still fails (as expected), but I don't see where the font rendering is wrong. Probably the font is even unused or maybe there is another font source for the same font family .

**Reply by ZuperZee** [reply]
I've made an issue <https://github.com/grafana/grafana/issues/32750> about it.
It comes from [the toolkit using a leading slash](<https://github.com/grafana/grafana/blob/82bae892eb030ea80428d0e4c271afaf59ad4539/packages/grafana-toolkit/src/config/webpack/loaders.ts#L163>).

**Reply by ZuperZee** [reply]
A quick fix would be to add
```
@import url('https://fonts.googleapis.com/css2?family=Open+Sans:wght@300;400;600;700&display=swap');
```
in the css, but it will require internet access to use the fonts.

![image](https://user-images.githubusercontent.com/39485579/114212293-f36fcd00-9961-11eb-990a-68bc9960511f.png)

**Reply by joerglemke** [reply]
Thanks for sharing.
Does it mean, these fonts are not required by the plugin backend editor, but just for the frontend rendering if referenced by the stylesheet?

**Reply by ZuperZee** [reply]
Yes. The fonts are fetched from the browser.
The reason why the OpenSans fonts were saved on the _backend_ is to make it available in a closed network.

The font is only for aesthetic. They aren't even used in the plugin. It has to be explicitly set in the CSS. Even then it's only for the elements in the panel htmlNode (HTML in the panel).

**Comment by ZuperZee** [comment, marked as answer]
Fixed in v1.3.2
(Cleaning up some discussions)
